qiskit/aqua/algorithms/single_sample/hhl/hhl.py

```diff
@@ -0,0 +1,328 @@
+        data = []
+        prep = None
+        for j, _ in enumerate(labels):
+            select_bitpos = None
+            for cbit_label in results.results[j].header.clbit_labels:
+                if cbit_label[0] == select.name:
+                    select_bitpos = cbit_label[1]
+            all_counts = results.get_counts(labels[j])
+            filt_counts = []
+            filt_keys = []
+            for k, v in all_counts.items():
+                if int(k[-1-select_bitpos]) == 1:
+                    filt_keys.append(k[:select_bitpos-2])
+                    filt_counts.append(v)
+            filt_labels = dict(zip(filt_keys, filt_counts))
+            if filt_labels == {}:
+                filt_labels = {'0': 0}
+            counts = tomo.marginal_counts(filt_labels, tomoset['qubits'])
+            shots = sum(counts.values())
+            meas = circuits[j]['meas']
+            prep = circuits[j].get('prep', None)
+            meas_qubits = sorted(meas.keys())
+            if prep:
+                prep_qubits = sorted(prep.keys())
+            circuit = {}
+            for c in counts.keys():
+                circuit[c] = {}
+                circuit[c]['meas'] = [(meas[meas_qubits[k]], int(c[-1 - k]))
+                                      for k in range(len(meas_qubits))]
+                if prep:
+                    circuit[c]['prep'] = [prep[prep_qubits[k]] for k in
+                                          range(len(prep_qubits))]
+            data.append({'counts': counts, 'shots': shots, 'circuit': circuit})
+        ret = {'data': data, 'meas_basis': tomoset['meas_basis']}
+        if prep:
+            ret['prep_basis'] = tomoset['prep_basis']
+        return ret
+
+    def _hhl_results(self, vec):
+        self._ret["output_hhl"] = vec
+        # Calculating the fidelity with the classical solution
+        theo = np.linalg.solve(self._matrix, self._vector)
+        theo = theo/np.linalg.norm(theo)
+        self._ret["fidelity_hhl_to_classical"] = abs(theo.dot(vec.conj()))**2
+        # Rescaling the output vector to the real solution vector
+        tmp_vec = self._matrix.dot(vec)
+        f1 = np.linalg.norm(self._vector)/np.linalg.norm(tmp_vec)
+        f2 = sum(np.angle(self._vector*tmp_vec.conj()-1+1))/self._num_q # "-1+1" to fix angle error for -0.-0.j
+        self._ret["solution_hhl"] = f1*vec*np.exp(-1j*f2)
+
+    def _run(self):
+        self.construct_circuit()
+        if self._quantum_instance.is_statevector:
+            self._statevector_simulation()
+        else:
+            self._state_tomography()
+        # Adding a bit of general result information
+        self._ret["input_matrix"] = self._matrix
+        self._ret["input_vector"] = self._vector
+        self._ret["eigenvalues_classical"] = np.linalg.eig(self._matrix)[0]
+        self._ret["solution_classical"] = list(np.linalg.solve(self._matrix, self._vector))
+        self._ret["circuit_width"] = self._circuit.width()
```

Comment:
minor efficiency issue, Terra converts the circuits to dag to get width and depth info. Thus, it will be more efficient to convert the circuit to dag once and then get `width` and `depth`.